### ISSUE TITLE
Terraform crashes when the aws "allowed_account_ids" uses the wrong account id.

### ISSUE THREAD

**Issue by josh-padnick** [hashicorp/terraform#8923]
I discovered an issue where, when Terraform attempts to update an AWS account whose ID doesn't match the one listed in the `allows_account_ids` property of `provider.aws`, instead of giving a graceful error message about the problem, Terraform crashes.
### Terraform Version

Terraform v0.7.3
### Affected Resource(s)
- provider.aws
- aws_iam_group
### Terraform Configuration Files

**main.tf**

``` hcl
provider "aws" {
  region = "${var.aws_region}"
  allowed_account_ids = ["${var.aws_account_id}"]
}

module "iam_group" {
  source = "./modules/iam-groups"
}
```

**modules/iam-group/main.tf**

``` hcl
resource "aws_iam_group" "admin" {
  name = "admin-temp"
}
```
### Debug Output

https://gist.github.com/josh-padnick/c6601137736758d63b84241be252c7ec
### Panic Output

https://gist.github.com/josh-padnick/a40d1a922684f8622b11b4ed0436cb26
### Expected Behavior

Terraform should have complained that I was attempting to update AWS Account A when my `provider.aws.allowed_account_ids` only permits updates to AWS Account B.
### Actual Behavior

Terraform crashed entirely.
### Steps to Reproduce
1. `terraform get` to load the `iam_group` module.
2. `terraform apply`
### Important Factoids

I've previously gotten a valid error message here, so I'm not sure where the regression came from. It would be interesting to find out if this is specific to the `aws_iam_group` resource, or generic to all AWS resources.

**Comment by rvandegrift** [comment]
Hit this issue today, but in a different context.  I'm trying to get terraform to work in the situation @josh-padnick reported.  (Maybe this is a feature request?)

I'm using multiple providers to deploy resources into two AWS accounts.  But my only access is via IAM roles.  So I can't supply a static key pair to the providers, and have to use env vars.  Below is a simplified version of my config.  It crashes on plan or apply, since the env vars can only have credentials for one of the providers.

If `-target` is used to specify a resource associated with the current environment credentials, terraform works fine.  With a real config, this is tedious and error-prone.  But it means some nearby features would get this kind of cross-account deployment working:

1. If the current account is not in a provider's `allowed_account_ids`, skip all resources associated with that provider.
2. Allow `-target` to take a provider alias.  Intepret this like `-target` was specified for each resource that depends on the specified provider.
3. Permit the aws provider to take `security_token`, so STS credentials could be passed in at runtime via a variable.

### Terraform Version
Terraform v0.7.3

### Affected Resource(s)
- `aws_*`
- `provider.aws`

### Terraform Configuration Files
```hcl
provider "aws" {
  alias = "alpha"
  region = "us-east-1"
  allowed_account_ids = ["${var.alpha}"]
}

provider "aws" {
  alias = "omega"
  region = "us-east-1"
  allowed_account_ids = ["${var.omega}"]
}

resource "aws_iam_role" "alpha-role" {
  provider = "aws.alpha"

  name = "alpha-role"
  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "Service": "lambda.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF
}

resource "aws_iam_role" "omega-role" {
  provider = "aws.omega"

  name = "omega-role"
  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "AWS": ["arn:aws:iam::${var.alpha}:alpha-role"]
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF
}
```

**Comment by josh-padnick** [comment]
@rvandegrift I haven't verified this myself, but it looks like https://github.com/hashicorp/terraform/pull/8731 was fixed and released in Terraform 0.7.4, so you might try upgrading!

**Comment by rvandegrift** [comment]
Thanks for the heads up - I don't see 0.7.4 on the download page yet, but it is there.  I just tested, the crash is fixed and replaced by an error message.  That confirms my use case is a feature request!
